pmlc/ast/builder.cc

```diff
@@ -752,6 +768,92 @@ struct ProgramBuilder {
         .result();
   }
 
+  Value handleLoop(ExprNodeLoop *node) {
+    // take lowBound, upperBound, step from operands.
+    auto indices = node->indexs;
+    auto indexType = builder.getIndexType();
+    std::vector<Value> indexValue;
+    for (auto index : indices) {
+      auto indexNum = builder.create<mlir::IndexCastOp>(
+          loc, builder.lookupNode(index), indexType);
+      indexValue.push_back(indexNum);
+    }
+
+    SmallVector<Value, 8> InitArgs;
+    for (const ExprNodePtr &operand : node->operands) {
+      InitArgs.push_back(builder.lookupNode(operand));
+    }
+    auto results = node->results;
+    DenseSet<Value> InitArgSet;
+    InitArgSet.insert(InitArgs.begin(), InitArgs.end());
+    SmallVector<Value, 8> resultArgs;
+    for (const ExprNodePtr &result : results) {
+      resultArgs.push_back(builder.lookupNode(result));
+    }
+    DenseSet<Value> resultArgSet;
+    resultArgSet.insert(resultArgs.begin(), resultArgs.end());
+
+    AstTraversal traversal(results);
+    auto scfForOp = builder.create<mlir::scf::ForOp>(
+        loc, indexValue[0], indexValue[1], indexValue[2], InitArgs);
+    OpBuilder bodyBuilder(scfForOp.getLoopBody());
+
+    BlockAndValueMapping mapper;
+    auto rawIterArgs = scfForOp.getRegionIterArgs();
+    for (auto tuple : llvm::zip(InitArgs, rawIterArgs)) {
+      Value outer, inner;
+      std::tie(outer, inner) = tuple;
+      mapper.map(outer, inner);
+    }
+    // choose the ops which should be put into loop body.
+    llvm::SetVector<Value> affectValue(InitArgs.begin(), InitArgs.end());
+    llvm::SetVector<Operation *> innerLoopValues;
+    while (!affectValue.empty()) {
+      auto uses = affectValue.back().getUses();
+      for (auto &use : uses) {
+        auto op = use.getOwner();
+        if (!innerLoopValues.contains(op) && op != scfForOp.getOperation()) {
+          for (auto result : op->getResults()) {
+            affectValue.insert(result);
+          }
+          innerLoopValues.insert(op);
+        }
+      }
+      affectValue.pop_back();
+    }
+
+    llvm::SetVector<Operation *> toRemove;
+    SmallVector<Value, 4> yieldValues;
+    for (const ExprNodePtr &node : traversal.getFlat()) {
+      Value value = builder.lookupNode(node);
+      // skip init value.
+      if (InitArgSet.count(value)) {
+        continue;
+      }
+      Operation *op = value.getDefiningOp();
+      // skip duplicate op, and not inside loop op.
+      if (!op || toRemove.contains(op) || !innerLoopValues.contains(op)) {
+        continue;
+      }
+      Operation *clonedOp = bodyBuilder.clone(*op, mapper);
+      if (resultArgSet.count(value)) {
```

Comment:
Please check if here has the same issue as Layer OP, see
https://github.com/plaidml/plaidml/pull/1708/files#diff-3f55e6bfdb58bd4443a3ba3a3da3c7ff34bdec545280a40d1959c3fd6b7d4826L723